JSONObject.java

```diff
@@ -1849,26 +1852,30 @@ static final Writer writeValue(Writer writer, Object value,
         } else if (value.getClass().isArray()) {
             new JSONArray(value).write(writer, indentFactor, indent);
         } else if (value instanceof Number) {
-            writer.write(numberToString((Number) value));
+            writer.append(numberToString((Number) value));
         } else if (value instanceof Boolean) {
-            writer.write(value.toString());
+            writer.append(value.toString());
         } else if (value instanceof JSONString) {
-            Object o;
+            String o;
             try {
                 o = ((JSONString) value).toJSONString();
             } catch (Exception e) {
                 throw new JSONException(e);
             }
-            writer.write(o != null ? o.toString() : quote(value.toString()));
+            if (o != null) {
+                writer.append(o);
+            } else {
+                quote(value.toString(), writer);
```

Comment:
This version of quote() writes directly to the Appendable, avoids another buffer copy.